db_stress_tool/db_stress_tool.cc

```diff
@@ -29,6 +29,8 @@
 
 namespace ROCKSDB_NAMESPACE {
 namespace {
+static std::shared_ptr<ROCKSDB_NAMESPACE::FileSystem> fs_guard;
+static std::shared_ptr<ROCKSDB_NAMESPACE::Env> composite_env_guard;
```

Comment:
I think `composite_env_guard` depends on `env_guard` for enforcing its dependencies outlive it when `--fs_uri` is provided. Then `env_guard` should be defined above `composite_env_guard` to ensure it is destroyed after.

But actually, this problem only arises because we have `composite_env_guard` as a non-local static. If instead it were a static local variable (maybe in a `GetCompositeEnv()` which could call `FileSystem::Load()` and `NewCompositeEnv()` on first invocation) , the proper destruction order should be enforced without any guards.

Reply:
Ping @anand1976  , since you did the refactoring of the env interface and the file system split, do you have time to comment on this? 

Thanks!

Reply:
Its not clear to me why there would be threads running with a dangling reference back to the composite env. The app should be responsible for cancelling/waiting for any such background jobs before destroying the composite env. IMO as @ajkr suggested, we should be able to do without ```GetCompositeBaseEnv``` and use a getter with a static variable instead of ```composite_env_guard```.

Reply:
I wanted to preserve the same behavior as for normal envs to avoid having new weird segfaults crop up when NewCompositeEnv would be used. What I saw(it i remember correctly) was that if something unexpected happens(like running out of disk space) with background work running (like compaction), a thread might continue running for some time with a reference back to the env it is operating in while the app(db_bench) is exiting. The only way(without changing the app code) is to safely destroy that reference is to destroy it's base env, joining all threads.  

Does that make sense as a background?

If the app is doing it wrong, then a getter with a static variable is much more straightforward and I can just fix any dangling-reference/unclean shutdown issues as I discover them.

Cheers!

Reply:
Understood the concern now. The way it's supposed to work in a clean shutdown is the app calls `DB::Close()` or delete `DB` on all its DBs. The DB shutdown process involves waiting for all its jobs in the background thread pool to finish (here, finish usually just means failing quickly due to noticing the `shutting_down_` flag is set). Following that, the thread pool should no longer have any opportunity to reference the composite `Env`.